### PR TITLE
Amazon service extensions

### PR DESCRIPTION
**Pull request by marcinczeczko**
I'd like to introduce a broader integration with AWS SDK v2. It consists of:
- Amazon services extensions root module
- Introduction of Amazon services common module to easily add support for other AWS clients.
- Refactored DynamoDB extension to be part of amazon services
- Added S3 client extension
- Quickstart guide that shows how to store and retrieve file using S3

More clients (SNS, SQS, etc.) to be added if the concept in this PR will be finalised and merged

### CODE REVIEW

**Comment by gsmet** [comment]
I cancelled the build, we are scarce on CI resources right now.

**Comment by gsmet** [comment]
Oh man, this is a large one :). It looks very interesting, I need to take more time to think about it and review it.

**Comment by marcinczeczko** [comment]
Yes I know it's quite large :) No worries, we are not in the rush.

**Comment by marcinczeczko** [comment]
@gsmet - Any thoughts on this PR ?

**Comment by gsmet** [comment]
I reviewed the Flyway PR yesterday, this one is next in line once I'm done with releasing 1.0.1.Final so you should have some feedback soon.

Sorry, the 1.0 work got me very late on reviewing large PRs.

**Comment by gsmet** [comment]
@marcinczeczko sorry about the delay. I'm starting to have a look just now.

**Review by gsmet — CHANGES_REQUESTED** [review]
So, in general, that's a big +1. I really like what you have here. Thanks for your work (and for your patience...).

It needs a rebase and some changes to the extension descriptor. The rest is mostly some small comments.

Also I'm not entirely sure about how your architected the configuration. I think we need to discuss it and it would probably be better to do it live if you can.

**Review by oscerd — APPROVED** [review]

**Comment by gsmet** [comment]
@marcinczeczko happy new year! Any chance we could restart the progress on this? It looked very promising!

**Comment by marcinczeczko** [comment]
@gsmet Happy new year too !! I just came back from long Christmas break :) I'm gonna to restart that work this week. I think we can talk about in live as you suggested - let's catchup on zulip to get up to speed.

File: bom/runtime/pom.xml
Comment (marcinczeczko) [reply bom/runtime/pom.xml]: Fixed

File: ci-templates/stages.yml
Comment (marcinczeczko) [reply ci-templates/stages.yml]: Only underscores are supported but doesn't looks good, so I kept camelCase with `amazonServices` as value

File: docs/src/main/asciidoc/amazon-s3-guide.adoc
Comment (marcinczeczko) [reply docs/src/main/asciidoc/amazon-s3-guide.adoc]: I forgot to mention, that S3 guide I'm about to work on later :) it's just a stub atm. Will polish it later this week.

File: extensions/amazon-services/common/deployment/src/main/java/io/quarkus/amazon/common/deployment/InterceptorsPathBuildItem.java
Comment (marcinczeczko) [reply extensions/amazon-services/common/deployment/src/main/java/io/quarkus/amazon/common/deployment/InterceptorsPathBuildItem.java]: fixed

File: integration-tests/amazon-services/pom.xml
Comment (marcinczeczko) [reply integration-tests/amazon-services/pom.xml]: fixed

File: extensions/amazon-services/s3/runtime/src/main/resources/META-INF/quarkus-extension.json
Comment (marcinczeczko) [reply extensions/amazon-services/s3/runtime/src/main/resources/META-INF/quarkus-extension.json]: fixed

File: extensions/amazon-services/dynamodb/runtime/src/main/resources/META-INF/quarkus-extension.json
Comment (marcinczeczko) [reply extensions/amazon-services/dynamodb/runtime/src/main/resources/META-INF/quarkus-extension.json]: fixed

File: extensions/amazon-services/common/runtime/src/main/resources/META-INF/quarkus-extension.json
Comment (marcinczeczko) [reply extensions/amazon-services/common/runtime/src/main/resources/META-INF/quarkus-extension.json]: fixed

**Comment by oztimpower** [comment]
Any progress on this one?  

It's super fiddly and time consuming trying to get basic AWS Java SDK v2 (or v1) integrations working. I have SNS working using the sync client UrlConnectionHttpClient yet couldn't get it working with Apache (SSL issues). 

Required reflect-config.json for the UrlConnectionHttpClient:
[
  {
  "name":"com.sun.xml.internal.stream.XMLInputFactoryImpl",
  "methods":[{"name":"<init>","parameterTypes":[] }]
  }
]

**Comment by sherl0cks** [comment]
Is there any plans to move this forward @marcinczeczko ?

@gsmet @oscerd  @oztimpower I must admit that I worry about the strategy being employed here, which seems to be taken from the initial dynamodb extension. Specifically, all the work being done to actually initiate clients and proxy the AWS SDK configuration model. It is seems unnecessarily complex and hard to maintain, and perhaps why this PR is stalled after months of works and 120 files edited.

As a user deploy service to production with S3, Dynamo, SNS and SQS with quarkus / graal native / aws lambda, all I want quarkus to do is to handle the ugly internals of the build system (e.g. proxy generation, reflection config, resources config, maven dependency management) such that when I create AWS SDK clients in my CDI configuration (which is easy to do), they just work. On occasion it might be necessary to do bytecode replacement to ensure effective integration with quarkus, as well. I don't need the beans created for me. You can see this alternative model in https://github.com/quarkusio/quarkus/tree/master/extensions/amazon-lambda-xray we worked on with @patriot1burke .

I have the above approach working in the applications I'm deploying now to production, and if there was interest in trying this new strategy, then we can discuss how best to build out these features together (along with a guide). I will probably need support from the maintainers give my existing job responsibilities. That said, if maintainers are set on the existing strategy, then I don't want to waste time implementing an alternative.

**Comment by oscerd** [comment]
> Is there any plans to move this forward @marcinczeczko ?
> 
> @gsmet @oscerd @oztimpower I must admit that I worry about the strategy being employed here, which seems to be taken from the initial dynamodb extension. Specifically, all the work being done to actually initiate clients and proxy the AWS SDK configuration model. It is seems unnecessarily complex and hard to maintain, and perhaps why this PR is stalled after months of works and 120 files edited.
> 
> As a user deploy service to production with S3, Dynamo, SNS and SQS with quarkus / graal native / aws lambda, all I want quarkus to do is to handle the ugly internals of the build system (e.g. proxy generation, reflection config, resources config, maven dependency management) such that when I create AWS SDK clients in my CDI configuration (which is easy to do), they just work. On occasion it might be necessary to do bytecode replacement to ensure effective integration with quarkus, as well. I don't need the beans created for me. You can see this alternative model in https://github.com/quarkusio/quarkus/tree/master/extensions/amazon-lambda-xray we worked on with @patriot1burke .
> 
> I have the above approach working in the applications I'm deploying now to production, and if there was interest in trying this new strategy, then we can discuss how best to build out these features together (along with a guide). I will probably need support from the maintainers give my existing job responsibilities. That said, if maintainers are set on the existing strategy, then I don't want to waste time implementing an alternative.

Hello, personally I'm working on supporting the AWS SDK v2 in the https://github.com/apache/camel-quarkus and I'm still trying to find a common way between the multiple components. I will post my findings here whenever I have some.

**Comment by sherl0cks** [comment]
@oscerd Interesting. I have some thoughts on this from my work and also a lot of experience with camel (I use to work with Red Hat consulting). I'll send you an email - perhaps we can discuss this over a video call quickly to coordinate?

**Comment by oscerd** [comment]
Sure thing!

**Comment by marcinczeczko** [comment]
Hi @oscard, yes I planned to resume that work. I had to stop for a while because of client work. I think we can discuss it in a wider group.

**Comment by sherl0cks** [comment]
@marcinczeczko sent email to your gmail to coordinate a discussion. Looking forward to discussing with you soon.

**Comment by marcinczeczko** [comment]
PR rebased

**Comment by oscerd** [comment]
Thanks

**Comment by gsmet** [comment]
Hi @marcinczeczko !

Sorry, things have been crazy these past months. That being said, I think I can commit to help you get this ready for Quarkus 1.5 now that things are a bit more normal.

We have a conflict right now due to this commit: https://github.com/quarkusio/quarkus/commit/f4252631a328386157d4c2e0be1c65a3771f42ac#diff-55e99caed6b5d071cd8d4111774de384 . I think we can safely rebase and ignore those changes for now.

As for my config comment, my question was if maybe having something like:
```
quarkus.amazon.dynamodb.endpoint-override=${dynamodb.url}
quarkus.amazon.dynamodb.interceptors=io.quarkus.it.amazon.dynamodb.DynamoDBModifyResponse

quarkus.amazon.dynamodb.region=eu-central-1
quarkus.amazon.dynamodb.credentials.type=static
quarkus.amazon.dynamodb.credentials.static-provider.access-key-id=test-key
quarkus.amazon.dynamodb.credentials.static-provider.secret-access-key=test-secret
```
e.g. without the `.aws.` map and all those configurations would be specific to a given service.

I'm not sure centralizing that config maps is such a good idea. Keep in mind that you can reuse the same config object in all configs.

That being said, you might have had a very good reason to do that so that's just a suggestion.

That's the only discussion we need to have before getting this PR in IMHO.

Thanks!

**Comment by marcinczeczko** [comment]
@gsmet - No worries. It was crazy on my side too.
Re your comments. I'll remove `.aws.` part as it was before that way and will rebase of course.

About the centralizing config maps. The reason why i did that is because all AWS clients share those bits and I wanted to have an ability to configure all of those aspects independently per client as one app can use multiple clients (e.g dynamo, s3, sns togeter, each have configured some of the aspects in a different way)

**Comment by gsmet** [comment]
> I wanted to have an ability to configure all of those aspects independently per client as one app can use multiple clients (e.g dynamo, s3, sns togeter, each have configured some of the aspects in a different way)

I'm not sure I understand why it wouldn't be possible with the approach I presented?

Except if you have the ability to configure one client and reuse it for several of the services. In this case your approach might make sense.

I don't know if it's possible to set up one client and use it for everything? And if it makes sense?

**Comment by marcinczeczko** [comment]
No no, I didn't meant to set up one client and use it for everything. I
meant to configure each client individually. E.g.

quarkus.amazon.dynamodb.interceptors=com.acme.Someinterceptor
quarkus.amazon.dynamodb.credentials.type=static

quarkus.amazon.s3.interceptors=com.acme.AnotherInterceptor
quarkus.amazon.dynamodb.credentials.type=process

And since the configurations like sdk, aws, sync/async-client are the same
no matter the client I'm binding them via the AmazonConfig in maps, so the
map key denotes a client name.
With this approach, next extensions for other clients such as SNS, SQS you
can start configure those clients too (quarkus.amazon.sns.....) And it's
centralized in common module as the buildsteps, http client generations is
the same but need to be executed per each client in order to instantiate
them. So if user of S3 would like to change some parameters in the HTTP
transport or enable some retries he can do it independently on the http
transport configured for dynamo, sns etc.

I think I might see what you mean in general, so for instance to be able to
instantiate one HTTP transport that can be reused by all Amazon clients -
that part I think might be a quite interesting scenario to consider too.

**Comment by gsmet** [comment]
> So if user of S3 would like to change some parameters in the HTTP transport or enable some retries he can do it independently on the http transport configured for dynamo, sns etc.

So I totally agree with this and I totally agree that it should be flexible.

IMHO we have 2 options:

1/ have the HTTP transport config in every service configuration. It could still be centralized and pushed via build items to a common build step that could instantiate everything. And use a common config object placed in the shared module to avoid duplication. I think build items would allow us to do that without too much hassle.
2/ do what you did.

If we do 1/ I think the config is a bit cleaner and better documented. It might be a bit harded to share the clients if we decide it makes sense. I think it could be doable though.

Not saying your approach is bad, it has its merits too. I think we need to carefully ponder what we want to do and how the config should be.

1/ has one other advantage: I think it would be easier to potentially allow to connect to several services e.g. several S3 resources for instance or several DynamoDB datasources. I think that could be useful. We could reuse the approach we have for the Agroal datasources relatively easily. Again, not something I would want for a first version of this patch, just food for thoughts for the general design.

As said, I'm not very versed in AWS so I push some thoughts to you as a base of discussion. I don't have a strong opinion about what would be best but I wanted to share this with you.

**Comment by marcinczeczko** [comment]
Option /1 make sense. I'll take look at this and get back to you

pt., 24 kwi 2020, 11:15 użytkownik Guillaume Smet <notifications@github.com>
napisał:

**Comment by marcinczeczko** [comment]
@gsmet - I was looking at option /1 - it took me some time as couple of refactors had to happen. However, I stuck with npes on configs passed to the recorders. I seems to me that since the common config objects that are needed to run recorders from common module are never initialised. The current PR contains without option /1, please take a look at the changes towards /1 - https://github.com/marcinczeczko/quarkus/commit/e9fc37f57257e32d7c3456e7bc2dc407031fa9cd
Here, I'm passing configs to the recorder: https://github.com/marcinczeczko/quarkus/commit/e9fc37f57257e32d7c3456e7bc2dc407031fa9cd#diff-15f838f2ce3c520e6529ee1be277a0f1R86
But the problem is that it's not the configRoot config object so it's alway null when app is running - that just my gut feeling

**Comment by marcinczeczko** [comment]
@gsmet - PR ready for final review.

**Comment by gsmet** [comment]
@machi1990 we have doc issues with this PR. I wonder if you could have a look?

There's something quite different here: I think it's the first time we use a shared config group that is in another extension.

I thought that maybe we could add the ability to point to other directories when analyzing the doc classes?

But you might have better ideas.

**Comment by machi1990** [comment]
> @machi1990 we have doc generation issues with this PR. I wonder if you could have a look?
> 
> There's something quite different here: I think it's the first time we use a shared config group that is in another extension.

@gsmet @marcinczeczko I have not gotten yet into checking what's happening here, but from what I gather, the shared config groups are in the common module  of their own and the config roots using them in the extension module, right? 

If so, I think the generator will need an enhancement to scan these and store them somewhere as I do not think they are scanned at all. 

> I thought that maybe we could add the ability to point to other directories when analyzing the doc classes?

Can you develop further?

**Comment by gsmet** [comment]
> I have not gotten yet into checking what's happening here, but from what I gather, the shared config groups are in the common module of their own and the config roots using them in the extension module, right?

Yes.

> Can you develop further?

I don't know how the config generator works, if it's extension per extension or if it has a global state. I was supposing the former and so that we might have to point the processor to the addiitional source files from the common code when working on S3 and DynamoDB.

**Comment by machi1990** [comment]
> > I have not gotten yet into checking what's happening here, but from what I gather, the shared config groups are in the common module of their own and the config roots using them in the extension module, right?
> 
> Yes.
> 
> > Can you develop further?
> 
> I don't know how the config generator works, if it's extension per extension or if it has a global state. 

Oh yes, it is extension per extension (say maven project per maven project) i.e a runtime module will be scanned separately, and the deployment module will be scanned separately, and we try to aggregate scanned configurations per extension (in all cases). 

> I was supposing the former and so that we might have to point the processor to the additional source files from the common code when working on S3 and DynamoDB.

This could have worked if config groups were scanned in all cases but at present we scan them depending on if they are used or not (inside a maven module that is currently being scanned, and this is driven by config roots). So this was pretty limited as can be seen now. 

All the scanning is happening here 
https://github.com/quarkusio/quarkus/blob/master/core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDoItemFinder.java#L62 

What I think we should do is scan config root first, and store them somewhere so that they can be referenced. 

I plan to have a look at this soon.

**Comment by gsmet** [comment]
OK, in the meantime, we can certainly make progress on the rest of the PR.

**Comment by marcinczeczko** [comment]
Opened PR for quickstart: https://github.com/quarkusio/quarkus-quickstarts/pull/550

**Comment by machi1990** [comment]
@marcinczeczko can you check if the PR https://github.com/quarkusio/quarkus/pull/8994 fixes the issue

**Comment by marcinczeczko** [comment]
@machi1990 - thanks, checked it and works perfect.

**Comment by marcinczeczko** [comment]
@gsmet - I rebased with master to pick up [#8994](https://github.com/quarkusio/quarkus/pull/8994). So documentation generates correctly now, I think you can proceed with the final review now.

**Review by gsmet — CHANGES_REQUESTED** [review]
@marcinczeczko I added some minor comments that needs fixing, could you take care of them and squash everything into a single commit?

There are a few things we need to discuss as I have a few questions but I suggest that we merge this one first and then we can have further discussions/improvements later on.

**Comment by gsmet** [comment]
Also, I saw we break the existing configuration properties. I suppose you find it better this way?

We can do that but we need to document it properly in the migration guide. We can do that later but prior to the release!

File: extensions/amazon-services/dynamodb/runtime/pom.xml
Comment (marcinczeczko) [reply extensions/amazon-services/dynamodb/runtime/pom.xml:14]: fixed

**Comment by marcinczeczko** [comment]
> Also, I saw we break the existing configuration properties. I suppose you find it better this way?
> 
> We can do that but we need to document it properly in the migration guide. We can do that later but prior to the release!

Do you mean configurations that had `.aws.`. If yes, I think there were no a strong reason. Since the `sdk` config properties are not prefixed I simply decided to do so for `aws` onces. If you think it's better to keep as it was before I can easily change it.

**Comment by gsmet** [comment]
I dunno. The SDK ones were purely technical. The AWS ones were more like "let's configure the AWS part" so it kinda makes sense to me.

If there's no good reason, I would say we could keep them as is and avoid breaking user configuration?

**Comment by marcinczeczko** [comment]
> I dunno. The SDK ones were purely technical. The AWS ones were more like "let's configure the AWS part" so it kinda makes sense to me.
> 
> If there's no good reason, I would say we could keep them as is and avoid breaking user configuration?

Agree & done

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
OK, let's wait for CI and merge :).

Thanks @marcinczeczko for this great work and your patience and tenacity!

**Comment by marcinczeczko** [comment]
Thanks. :)

So, next in a row are SQS and SNS extensions, so your patience will be
required too.

śr., 6 maj 2020, 17:11 użytkownik Guillaume Smet <notifications@github.com>
napisał:

**Comment by sherl0cks** [comment]
@marcinczeczko Thanks for making this happen!

**Comment by oscerd** [comment]
Thanks @marcinczeczko

**Comment by gsmet** [comment]
Cool, looking forward to it. Now that the basics are covered, it should be easier.

**Comment by hrstoyanov** [comment]
@marcinczeczko thanks for the effort! Any chance we can get SES (simple email service)

**Comment by marcinczeczko** [comment]
@hrstoyanov - thanks :) Actually I just opened PR #9413 with SES (and KMS)

**Pull request merged**
